BJS2-40651 Project Service : Vacancies - Filter Vacancies Feature

drewbal opened the pull request:
4/5 of Vacancy Feature

drewbal committed:
Initial commit

drewbal committed:
BJS2-40651 Feature: Create Vacancy - adds VacancyDto

DTO: VacancyDto added

drewbal committed:
BJS2-40651 Feature: Create Vacancy - adds Vacancy Mapper

DTO: VacancyDto optimized
Mapper: VacancyMapper added

drewbal committed:
BJS2-40651 Feature: Create Vacancy - adds controller and create method

VacancyController: createVacancy method added

drewbal committed:
BJS2-40651 Feature: Create Vacancy - create full vacancy from dto

VacancyService: toEntityFromDto method added (add fields ignored by mapper)
ProjectService: getProjectById metod added

drewbal committed:
BJS2-40651 Feature: Create Vacancy - validate data to create the vacancy

VacancyValidator: checks the project is valid; role of the creator is valid;
TeamMemberService: returns the team member by id
DataValidationException: new Exception

drewbal committed:
BJS2-40651 Feature: Create Vacancy - creates new vacancy in system

VacancyService: creates a new vacancy;

drewbal committed:
BJS2-40651 Feature: Create Vacancy - sets status at creation

VacancyDto: VacationStatus field added;
VacancyService: sets OPEN status at creation;

drewbal committed:
BJS2-40651 Feature: Create Vacancy - review fixes

VacancyService: toEntityFromDto is private;
VacancyController: path fixed;
Other: blank lines removed

drewbal committed:
BJS2-40651 Feature: Update Vacancy - controller

VacancyController: new method to update the existing vacancy status

drewbal committed:
BJS2-40651 Feature: Update Vacancy - dto update

VacancyDto: upgradeBy field added
Other: VacancyMapper class moved to other package

drewbal committed:
BJS2-40651 Feature: Update Vacancy - validator update

VacancyValidator: checks the role od updater and the number of candidates for vacancy closure
Exception: InsufficientCandidatesException added

drewbal committed:
BJS2-40651 Feature: Update Vacancy - can update vacancy status to CLOSED

VacancyService: updateVacancyStatus method that changes the status of the vacancy based on candidates amount criteria

drewbal committed:
Merge branch 'Dragon-master-stream7' into BJS2-40651-update-vacancy

drewbal committed:
BJS2-40651 Feature: Update Vacancy - review fixes

VacancyController: api v1 added to path
VacancyService: @Transactional annotation added

drewbal committed:
BJS2-40651 Feature: Delete Vacancy - controller upgrade

VacancyController: delete vacancy method added

drewbal committed:
BJS2-40651 Feature: Delete Vacancy - service upgrade

VacancyService: delete vacancy method added
CandidateService: find candidate by id added

drewbal committed:
BJS2-40651 Feature: Create Vacancy - review fixes

VacancyValidator: logic changed, VacancyRepository injected
VacancyService: validation fix
ProjectValidator: basic validation added

drewbal committed:
Merge branch 'BJS2-40651-create-vacancy' into BJS2-40651-update-vacancy

# Conflicts:
#	src/main/java/faang/school/projectservice/service/VacancyService.java
#	src/main/java/faang/school/projectservice/validator/VacancyValidator.java
#	src/test/java/faang/school/projectservice/validator/VacancyValidatorTest.java

drewbal committed:
BJS2-40651 Feature: Update Vacancy - review fixes

Other: minor conflict fixes

drewbal committed:
Merge branch 'BJS2-40651-update-vacancy' into BJS2-40651-delete-vacancy

drewbal committed:
BJS2-40651 Feature: Delete Vacancy - deletes vacancy by id

VacancyService: deleteVacancy method added

drewbal committed:
BJS2-40651 Feature: Filter Vacancies - controller upgrade

VacancyController: Filter vacancies method added
FilterVacancyDto: added with 3 filters

drewbal committed:
BJS2-40651 Feature: Filter Vacancies - adds filters

VacancyFilters: Vacancy Title, Salary and WorkSchedule filters added to the application.

drewbal committed:
BJS2-40651 Feature: Filter Vacancies - service upgrade

VacancyService: FilterVacancies method with expansion ability through list of new filters

drewbal committed:
Merge branch 'Dragon-master-stream7' into BJS2-40651-create-vacancy

# Conflicts:
#	src/main/java/faang/school/projectservice/exception/DataValidationException.java
#	src/main/java/faang/school/projectservice/service/TeamMemberService.java

drewbal committed:
BJS2-40651 Feature: Create Vacancy - merge and conflicts solving

Minor fixes

drewbal committed:
Merge branch 'BJS2-40651-create-vacancy' into BJS2-40651-update-vacancy

# Conflicts:
#	src/main/java/faang/school/projectservice/mapper/VacancyMapper.java
#	src/main/java/faang/school/projectservice/validator/VacancyValidator.java
#	src/test/java/faang/school/projectservice/controller/VacancyControllerTest.java
#	src/test/java/faang/school/projectservice/validator/VacancyValidatorTest.java

DeemaP commented on src/main/java/faang/school/projectservice/dto/vacancy/FilterVacancyDto.java:
```diff
@@ -0,0 +1,23 @@
+
+@Data
+@Builder
+public class FilterVacancyDto {
```
Если это фильтр дто, то тут ведь не стоит вешать @NotNull, ведь на null значении у тебя построен принцип проверки фильтра isApplicable) А тут ты заведомо требуешь от людей фильтровать по этим значениям

drewbal replied:
Точно! Вот спасибо! Гениальный косяк :)))

drewbal committed:
BJS2-40651 Feature: Update Vacancy - merge and conflicts solving

Minor fixes

drewbal committed:
Merge branch 'BJS2-40651-update-vacancy' into BJS2-40651-delete-vacancy

drewbal committed:
Merge branch 'BJS2-40651-delete-vacancy' into BJS2-40651-filter-vacancy

DeemaP commented on src/main/java/faang/school/projectservice/service/ProjectService.java:
```diff
@@ -0,0 +1,20 @@
+        return projectRepository.getProjectById(projectId);
+    }
+
+    public boolean checkProjectExistsById(Long projectId) {
```
Так все-таки что это за валидация в сервисе? Или ты нагоняя от ментора ждешь?)
 А ведь он нигде не используется у тебя
 да еще и тесты написаны =)

drewbal replied:
Да, это пережиток нашей дискуссии. Прописал эту логику в валидаторе, а отсюда забыл убрать

DeemaP commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,91 @@
+                .toList();
+    }
+
+    public List<Candidate> getCandidatesByVacancyId(Long vacancyId) {
```
Почему public? Я вижу он используется только внутри этого сервиса и тестах.
Сделай private и тесты отдельные будут не нужны)

drewbal replied:
Ну вот этот наверное можно подумать и сделать приватным. Я просто не знаю, как у нас дальнейшая архитектура будет развиваться. Список кандидатов ведь может где-то использоваться в части hr управления проектами.

DeemaP commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,91 @@
+        return getVacancyById(vacancyId).getCandidates();
+    }
+
+    public Vacancy getVacancyById(long vacancyId) {
```
Так же и с этим методом, тем более на него даже тестов нет, хотя он public)

drewbal replied:
На него отдельный тест не нужен, т.к. он покрыт проверкой через другие методы. 

По поводу доступа. Вот представь, понадобится вдруг кому внутри проекта вакансия (или список кандидатов из предыдущего коммента). Как он его достанет, если тут табличка: "Дима запретил". Ну тогда к тебе и пойдут, как к хранителю центральной сущности.

Я считаю, что приватными нужно делать вспомогательный утилитарные методы, а те, которые достают общую инфу из бд надо оставить публичными

DeemaP replied:
По такой логике у тебя ниже есть метод getRejectedCandidatesIds, его тоже паблик надо делать, ведь он общую инфу из бд достает.

Надо тогда все в одной парадигме писать, либо так, либо так. 

А то получается, что getCandidatesByVacancyId никому не понадобится, а getVacancyById понадобится.

drewbal replied:
А вот смотри. 
`getCandidatesByVacancyId` вроде как из названия можно судить, что достает кандидатов. Просто кандидатов, общую инфу.
`getRejectedCandidatesIds` достает каких-то rejected кандидатов. Rejected кем? Почему? За что? То есть тут уже зашита какая-то логика валидации, а значит инфа уже обработанная и не полная. Поэтому метод служебный, приватный.

Если мы закроем все в сервисах, то не сможем выгрузить ни одной сущности для работы. Закрой все методы проектов у себя, вот весело будет. Но если все сделать публичным, то свалимся в хаос, некоторые методы будут вызываться там, где не должны. 

Везде нужен баланс, имхо, нет строгого либо так, либо так

DeemaP replied:
Ну тогда делай метод выше приватным, а этот, раз он тебе сущность возвращает, оставляй паблик, вдруг и правдва кому-то будет нужна сама вакансия. Остальные данные из нее уже сами вытащат))

drewbal replied:
Да, это как раз логичное решение, если можно вытащить инфу из сущности. И методов наружу меньше. Супер!

DeemaP reviewed (commented):
На тестах уже глаза поплыли, так что не осилил. Но по коду замечания ерундовые) Красиво)

drewbal committed:
BJS2-40651 Feature: Filter Vacancy: review fixes

VacancyService: incapsulation improved by making some methods private
FilterVacancyDTO: validation annotation refactored
ProjectService: unneccessary method removed

drewbal committed:
BJS2-40651 Feature: Create Vacancy: review fixes

VacancyController: logging added
ServiceController: logging added
GlobalExceptionHandler: fixed wrong import, logging added
Other: spaces fixed

drewbal committed:
Merge branch 'BJS2-40651-create-vacancy' into BJS2-40651-update-vacancy

# Conflicts:
#	src/main/java/faang/school/projectservice/service/VacancyService.java

drewbal committed:
BJS2-40651 Feature: Create Vacancy: review fixes

VacancyController: logging added
ServiceController: logging added
GlobalExceptionHandler: InsufficientCandidates exception added
Other: Exception imports fixed

drewbal committed:
Merge branch 'BJS2-40651-update-vacancy' into BJS2-40651-delete-vacancy

drewbal committed:
BJS2-40651 Feature: Delete Vacancy: review fixes

VacancyController: logging added

drewbal committed:
Merge branch 'BJS2-40651-delete-vacancy' into BJS2-40651-filter-vacancy

# Conflicts:
#	src/main/java/faang/school/projectservice/service/VacancyService.java

drewbal committed:
BJS2-40651 Feature: Filter Vacancy: review fixes

VacancyController: logging added
VacancyService: @Transactional from create method removed

drewbal committed:
BJS2-40651 Feature: Create Vacancy: optimization

NewVacancyDto: new DTO for vacancy creation
VacancyController: createVacancy method accepts new dto with minimum required fields
VacancyMapper: takes new vacancy DTO, returns default DTO
VacancyService: create method optimized for new dto
VacancyValidator: role validation methods united

drewbal committed:
Merge branch 'BJS2-40651-create-vacancy' into BJS2-40651-update-vacancy

# Conflicts:
#	src/main/java/faang/school/projectservice/controller/VacancyController.java
#	src/main/java/faang/school/projectservice/mapper/VacancyMapper.java
#	src/main/java/faang/school/projectservice/service/VacancyService.java

drewbal committed:
BJS2-40651 Feature: Update Vacancy: optimization

VacancyUpdateDto: new DTO for vacancy update
VacancyController: updateVacancy method accepts new dto with minimum required fields
VacancyService: update method optimized for new dto
VacancyValidator: obsolete role validation method removed

drewbal committed:
Merge branch 'BJS2-40651-update-vacancy' into BJS2-40651-delete-vacancy

# Conflicts:
#	src/test/java/faang/school/projectservice/service/VacancyServiceTest.java

drewbal committed:
BJS2-40651 Feature: Delete Vacancy: minor fixes

Tests fixes

drewbal committed:
Merge branch 'BJS2-40651-delete-vacancy' into BJS2-40651-filter-vacancy

# Conflicts:
#	src/main/java/faang/school/projectservice/controller/VacancyController.java
#	src/main/java/faang/school/projectservice/service/VacancyService.java
#	src/test/java/faang/school/projectservice/controller/VacancyControllerTest.java
#	src/test/java/faang/school/projectservice/service/VacancyServiceTest.java

drewbal committed:
BJS2-40651 Feature: Delete Vacancy: minor fixes

Tests fixes

drewbal committed:
BJS2-40651 Feature: Create Vacancy: review fixes

Vacancy: createdBy -> createdById, updatedBy ->updatedById
NewVacancyDto: validation annotations fixed
VacancyDto: renamed in VacancyResponseDto to improve logic
Other: exception messages fixed

drewbal committed:
Merge branch 'BJS2-40651-create-vacancy' into BJS2-40651-update-vacancy

# Conflicts:
#	src/main/java/faang/school/projectservice/controller/VacancyController.java
#	src/main/java/faang/school/projectservice/dto/vacancy/VacancyDto.java
#	src/main/java/faang/school/projectservice/service/VacancyService.java
#	src/test/java/faang/school/projectservice/controller/VacancyControllerTest.java
#	src/test/java/faang/school/projectservice/service/VacancyServiceTest.java
#	src/test/java/faang/school/projectservice/validator/VacancyValidatorTest.java

drewbal committed:
BJS2-40651 Feature: Update Vacancy: review fixes

Fixes from merge of Create Vacancy Feature

drewbal committed:
Merge branch 'BJS2-40651-update-vacancy' into BJS2-40651-delete-vacancy

# Conflicts:
#	src/test/java/faang/school/projectservice/controller/VacancyControllerTest.java
#	src/test/java/faang/school/projectservice/service/VacancyServiceTest.java

drewbal committed:
BJS2-40651 Feature: Delete Vacancy: review fixes

Fixes from merge of Update Vacancy Feature

drewbal committed:
Merge branch 'BJS2-40651-delete-vacancy' into BJS2-40651-filter-vacancy

# Conflicts:
#	src/main/java/faang/school/projectservice/controller/VacancyController.java
#	src/main/java/faang/school/projectservice/service/VacancyService.java
#	src/test/java/faang/school/projectservice/controller/VacancyControllerTest.java
#	src/test/java/faang/school/projectservice/service/ProjectServiceTest.java
#	src/test/java/faang/school/projectservice/service/VacancyServiceTest.java

drewbal committed:
BJS2-40651 Feature: Filter Vacancy: review fixes

Fixes from merge of Delete Vacancy Feature
